File: sdk/go/captcha.go
package captcha

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"crypto/sha512"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strconv"
	"strings"
	"time"
)

type SignatureAlgorithm string

const (
	AlgorithmHMACSHA256 SignatureAlgorithm = "HMAC-SHA256"
	AlgorithmHMACSHA512 SignatureAlgorithm = "HMAC-SHA512"
	AlgorithmBlake2b256 SignatureAlgorithm = "BLAKE2B-256"
	AlgorithmBlake2b512 SignatureAlgorithm = "BLAKE2B-512"
)

type Client struct {
	baseURL             string
	httpClient          *http.Client
	apiKey              string
	signatureKey        string
	signatureAlgorithm  SignatureAlgorithm
	enableKeyRotation   bool
}

func NewClient(baseURL string, options ...Option) *Client {
	c := &Client{
		baseURL: baseURL,
		httpClient: &http.Client{
			Timeout: 30 * time.Second,
		},
		signatureAlgorithm: AlgorithmHMACSHA256,
	}

	for _, opt := range options {
		opt(c)
	}

	return c
}

type Option func(*Client)

func WithHTTPClient(client *http.Client) Option {
	return func(c *Client) {
		c.httpClient = client
	}
}

func WithAPIKey(apiKey string) Option {
	return func(c *Client) {
		c.apiKey = apiKey
	}
}

func WithTimeout(timeout time.Duration) Option {
	return func(c *Client) {
		c.httpClient.Timeout = timeout
	}
}

func WithSignatureKey(key string) Option {
	return func(c *Client) {
		c.signatureKey = key
	}
}

func WithSignatureAlgorithm(algorithm SignatureAlgorithm) Option {
	return func(c *Client) {
		c.signatureAlgorithm = algorithm
	}
}

func WithKeyRotation(enable bool) Option {
	return func(c *Client) {
		c.enableKeyRotation = enable
	}
}

func (c *Client) generateNonce(length int) (string, error) {
	bytes := make([]byte, length)
	if _, err := io.ReadFull(c.randReader(), bytes); err != nil {
		return "", err
	}
	return hex.EncodeToString(bytes)[:length], nil
}

func (c *Client) randReader() io.Reader {
	return bytes.NewReader(make([]byte, 32))
}

func (c *Client) sortQueryString(query string) string {
	if query == "" {
		return ""
	}

	values := make(map[string][]string)
	parts := strings.Split(query, "&")

	for _, part := range parts {
		kv := strings.SplitN(part, "=", 2)
		key := kv[0]
		value := ""
		if len(kv) > 1 {
			value = kv[1]
		}
		values[key] = append(values[key], value)
	}

	var keys []string
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	var result []string
	for _, key := range keys {
		for _, value := range values[key] {
			result = append(result, key+"="+value)
		}
	}

	return strings.Join(result, "&")
}

func (c *Client) hashBody(body []byte) string {
	if len(body) == 0 {
		return ""
	}
	h := sha256.Sum256(body)
	return hex.EncodeToString(h[:])
}

func (c *Client) buildStringToSign(method, path, query string, timestamp int64, nonce, bodyHash string) string {
	var parts []string
	parts = append(parts, strings.ToUpper(method))
	parts = append(parts, path)

	if query != "" {
		sortedQuery := c.sortQueryString(query)
		parts = append(parts, sortedQuery)
	}

	parts = append(parts, strconv.FormatInt(timestamp, 10))

	if nonce != "" {
		parts = append(parts, nonce)
	}

	if bodyHash != "" {
		parts = append(parts, bodyHash)
	}

	return strings.Join(parts, "\n")
}

func (c *Client) computeSignature(key, data string) string {
	switch c.signatureAlgorithm {
	case AlgorithmHMACSHA256:
		mac := hmac.New(sha256.New, []byte(key))
		mac.Write([]byte(data))
		return hex.EncodeToString(mac.Sum(nil))

	case AlgorithmHMACSHA512:
		mac := hmac.New(sha512.New, []byte(key))
		mac.Write([]byte(data))
		return hex.EncodeToString(mac.Sum(nil))

	case AlgorithmBlake2b256, AlgorithmBlake2b512:
		h := sha512.New()
		h.Write([]byte(key))
		h.Write([]byte(data))
		hash := h.Sum(nil)
		if c.signatureAlgorithm == AlgorithmBlake2b256 {
			return hex.EncodeToString(hash[:32])
		}
		return hex.EncodeToString(hash)

	default:
		mac := hmac.New(sha256.New, []byte(key))
		mac.Write([]byte(data))
		return hex.EncodeToString(mac.Sum(nil))
	}
}

func (c *Client) generateSignature(method, path, query string, timestamp int64, nonce string, body []byte) (string, error) {
	if c.signatureKey == "" {
		return "", fmt.Errorf("signature key not configured")
	}

	bodyHash := c.hashBody(body)
	stringToSign := c.buildStringToSign(method, path, query, timestamp, nonce, bodyHash)
	return c.computeSignature(c.signatureKey, stringToSign), nil
}

func (c *Client) addSignatureHeaders(req *http.Request, body []byte) error {
	if c.signatureKey == "" {
		return nil
	}

	timestamp := time.Now().Unix()
	nonce, err := c.generateNonce(16)
	if err != nil {
		return fmt.Errorf("failed to generate nonce: %w", err)
	}

	path := req.URL.Path
	query := req.URL.RawQuery

	signature, err := c.generateSignature(req.Method, path, query, timestamp, nonce, body)
	if err != nil {
		return fmt.Errorf("failed to generate signature: %w", err)
	}

	req.Header.Set("X-Timestamp", strconv.FormatInt(timestamp, 10))
	req.Header.Set("X-Nonce", nonce)
	req.Header.Set("X-Signature", signature)
	req.Header.Set("X-Signature-Algorithm", string(c.signatureAlgorithm))

	return nil
}

// SliderCaptchaResponse 滑块验证码响应
type SliderCaptchaResponse struct {
	SessionID    string `json:"session_id"`
	ImageURL    string `json:"image_url"`
	PuzzleURL   string `json:"puzzle_url"`
	HintURL     string `json:"hint_url"`
	Shape       int    `json:"shape"`
	SecretY     int    `json:"secret_y"`
	ImageWidth  int    `json:"image_width"`
	ImageHeight int    `json:"image_height"`
}

// TrajectoryPoint 轨迹点
type TrajectoryPoint struct {
	X int   `json:"x"`
	Y int   `json:"y"`
	T int64 `json:"t"`
}

// VerifyCaptchaRequest 验证码验证请求
type VerifyCaptchaRequest struct {
	SessionID string             `json:"session_id"`
	X         int                `json:"x"`
	Y         int                `json:"y,omitempty"`
	Trajectory []TrajectoryPoint `json:"trajectory,omitempty"`
}

// VerifyCaptchaResponse 验证码验证响应
type VerifyCaptchaResponse struct {
	Success        bool           `json:"success"`
	Message      string         `json:"message"`
	Remaining    int              `json:"remaining_attempts"`
	TrajectoryResult *TrajectoryResult `json:"trajectory_result,omitempty"`
}

// TrajectoryResult 轨迹分析结果
type TrajectoryResult struct {
	Score   float64  `json:"score"`
	Passed  bool     `json:"passed"`
	Reasons []string `json:"reasons,omitempty"`
}

// GetSliderCaptcha 获取滑块验证码
func (c *Client) GetSliderCaptcha(width, height, tolerance int) (*SliderCaptchaResponse, error) {
	url := fmt.Sprintf("%s/api/v1/captcha/slider", c.baseURL)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	q := req.URL.Query()
	if width > 0 {
		q.Add("width", fmt.Sprintf("%d", width))
	}
	if height > 0 {
		q.Add("height", fmt.Sprintf("%d", height))
	}
	if tolerance > 0 {
		q.Add("tolerance", fmt.Sprintf("%d", tolerance))
	}
	req.URL.RawQuery = q.Encode()

	if c.apiKey != "" {
		req.Header.Set("X-API-Key", c.apiKey)
	}

	if err := c.addSignatureHeaders(req, nil); err != nil {
		return nil, err
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result struct {
		Code    int                  `json:"code"`
		Message string               `json:"message"`
		Data    SliderCaptchaResponse `json:"data"`
	}

	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if result.Code != 0 {
		return nil, fmt.Errorf("API error: %s", result.Message)
	}

	return &result.Data, nil
}

// VerifyCaptcha 验证验证码
func (c *Client) VerifyCaptcha(req *VerifyCaptchaRequest) (*VerifyCaptchaResponse, error) {
	url := fmt.Sprintf("%s/api/v1/captcha/verify", c.baseURL)

	reqBody, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(reqBody))
	if err != nil {
		return nil, err
	}

	httpReq.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		httpReq.Header.Set("X-API-Key", c.apiKey)
	}

	if err := c.addSignatureHeaders(httpReq, reqBody); err != nil {
		return nil, err
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result struct {
		Code    int                  `json:"code"`
		Message string               `json:"message"`
		Data    VerifyCaptchaResponse `json:"data"`
	}

	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if result.Code != 0 {
		return nil, fmt.Errorf("API error: %s", result.Message)
	}

	return &result.Data, nil
}

// UserAuth 用户认证相关
type UserAuth struct {
	client *Client
}

// Auth 获取用户认证
func (c *Client) Auth() *UserAuth {
	return &UserAuth{client: c}
}

// LoginRequest 登录请求
type LoginRequest struct {
	Username     string `json:"username"`
	Password     string `json:"password"`
	CaptchaToken string `json:"captcha_token,omitempty"`
}

// LoginResponse 登录响应
type LoginResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
	ExpiresIn    int64  `json:"expires_in"`
	User         struct {
		ID       uint   `json:"id"`
		Username string `json:"username"`
		Email    string `json:"email"`
	} `json:"user"`
}

// Login 用户登录
func (u *UserAuth) Login(req *LoginRequest) (*LoginResponse, error) {
	url := fmt.Sprintf("%s/api/v1/auth/login", u.client.baseURL)

	reqBody, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	httpReq, err := http.NewRequest("POST", url, bytes.NewBuffer(reqBody))
	if err != nil {
		return nil, err
	}

	httpReq.Header.Set("Content-Type", "application/json")
	if u.client.apiKey != "" {
		httpReq.Header.Set("X-API-Key", u.client.apiKey)
	}

	if err := u.client.addSignatureHeaders(httpReq, reqBody); err != nil {
		return nil, err
	}

	resp, err := u.client.httpClient.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var result struct {
		Code    int          `json:"code"`
		Message string       `json:"message"`
		Data    LoginResponse `json:"data"`
	}

	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}

	if result.Code != 0 {
		return nil, fmt.Errorf("API error: %s", result.Message)
	}

	return &result.Data, nil
}

// Environment 环境检测相关
type Environment struct {
	client *Client
}

// Env 获取环境检测
func (c *Client) Env() *Environment {
	return &Environment{client: c}
}

// DetectionScriptResponse 环境检测响应
type DetectionScriptResponse struct {
	Script string
}

// GetDetectionScript 获取检测脚本
func (e *Environment) GetDetectionScript(callback string) (string, error) {
	url := fmt.Sprintf("%s/api/v1/detect/script", e.client.baseURL)

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", err
	}

	if callback != "" {
		q := req.URL.Query()
		q.Add("callback", callback)
		req.URL.RawQuery = q.Encode()
	}

	if e.client.apiKey != "" {
		req.Header.Set("X-API-Key", e.client.apiKey)
	}

	resp, err := e.client.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	return string(body), nil
}

// Usage example
func ExampleClient() {
	// 创建客户端
	client := NewClient("http://localhost:8080")

	// 获取滑块验证码
	captcha, err := client.GetSliderCaptcha(320, 160, 8)
	if err != nil {
		panic(err)
	}

	fmt.Printf("Session ID: %s\n", captcha.SessionID)

	// 验证验证码（这里假设用户滑动到了x=185的位置
	verifyReq := &VerifyCaptchaRequest{
		SessionID: captcha.SessionID,
		X:         185,
		Y:         captcha.SecretY,
		Trajectory: []TrajectoryPoint{
			{X: 0, Y: captcha.SecretY, T: time.Now().UnixMilli() - 1000},
			{X: 50, Y: captcha.SecretY + 5, T: time.Now().UnixMilli() - 800},
			{X: 100, Y: captcha.SecretY - 3, T: time.Now().UnixMilli() - 500},
			{X: 150, Y: captcha.SecretY + 2, T: time.Now().UnixMilli() - 200},
			{X: 185, Y: captcha.SecretY, T: time.Now().UnixMilli()},
		},
	}

	result, err := client.VerifyCaptcha(verifyReq)
	if err != nil {
		panic(err)
	}

	fmt.Printf("Success: %v, Message: %s\n", result.Success, result.Message)
}

File: sdk/python/captcha.py
"""
行为验证系统 Python SDK
"""

import requests
import json
import time
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, asdict


@dataclass
class TrajectoryPoint:
    """轨迹点"""
    x: int
    y: int
    t: int


@dataclass
class SliderCaptchaResponse:
    """滑块验证码响应"""
    session_id: str
    image_url: str
    puzzle_url: str
    hint_url: str
    shape: int
    secret_y: int
    image_width: int
    image_height: int


@dataclass
class VerifyCaptchaResponse:
    """验证码验证响应"""
    success: bool
    message: str
    remaining_attempts: int
    trajectory_result: Optional[Dict] = None


@dataclass
class LoginResponse:
    """登录响应"""
    access_token: str
    refresh_token: str
    expires_in: int
    user: Dict


class CaptchaClient:
    """验证码客户端"""
    
    def __init__(
        self, base_url: str, api_key: Optional[str] = None, timeout: int = 30):
        """
        初始化客户端
        
        Args:
            base_url: API基础URL
            api_key: API密钥
            timeout: 请求超时时间（秒）
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()
        self._token = None
        self._refresh_token = None
    
    def _get_headers(self) -> Dict[str, str]:
        """获取请求头"""
        headers = {
            'Content-Type': 'application/json',
        }
        if self.api_key:
            headers['X-API-Key'] = self.api_key
        if self._token:
            headers['Authorization'] = f'Bearer {self._token}'
        return headers
    
    def _request(
        self,
        method: str,
        path: str,
        data: Optional[Dict] = None,
        params: Optional[Dict] = None
    ) -> Any:
        """
        发送请求
        
        Args:
            method: HTTP方法
            path: API路径
            data: 请求数据
            params: URL参数
            
        Returns:
            响应数据
        """
        url = f"{self.base_url}{path}"
        
        kwargs = {
            'headers': self._get_headers(),
            'timeout': self.timeout,
        }
        
        if data:
            kwargs['json'] = data
        if params:
            kwargs['params'] = params
        
        response = self.session.request(method, url, **kwargs)
        response.raise_for_status()
        
        result = response.json()
        
        if result.get('code') != 0:
            raise Exception(f"API Error: {result.get('message', 'Unknown error')}")
        
        return result.get('data')
    
    def get_slider_captcha(
        self,
        width: int = 320,
        height: int = 160,
        tolerance: int = 8
    ) -> SliderCaptchaResponse:
        """
        获取滑块验证码
        
        Args:
            width: 图片宽度
            height: 图片高度
            tolerance: 容差值
            
        Returns:
            滑块验证码响应
        """
        data = self._request('GET', '/api/v1/captcha/slider', params={
            'width': width,
            'height': height,
            'tolerance': tolerance,
        })
        return SliderCaptchaResponse(**data)
    
    def verify_captcha(
        self,
        session_id: str,
        x: int,
        y: Optional[int] = None,
        trajectory: Optional[List[TrajectoryPoint]] = None
    ) -> VerifyCaptchaResponse:
        """
        验证滑块验证码
        
        Args:
            session_id: 会话ID
            x: X坐标
            y: Y坐标
            trajectory: 轨迹数据
            
        Returns:
            验证响应
        """
        req_data = {'session_id': session_id, 'x': x}
        if y is not None:
            req_data['y'] = y
        if trajectory:
            req_data['trajectory'] = [asdict(p) for p in trajectory]
        
        data = self._request('POST', '/api/v1/captcha/verify', data=req_data)
        return VerifyCaptchaResponse(**data)
    
    def get_click_captcha(self) -> Dict:
        """获取点击验证码"""
        return self._request('GET', '/api/v1/captcha/click')
    
    def get_gesture_captcha(self) -> Dict:
        """获取手势验证码"""
        return self._request('GET', '/api/v1/captcha/gesture')
    
    def verify_gesture_captcha(self, session_id: str, pattern: List[int]) -> Dict:
        """
        验证手势验证码
        
        Args:
            session_id: 会话ID
            pattern: 手势模式
            
        Returns:
            验证响应
        """
        return self._request('POST', '/api/v1/captcha/gesture/verify', data={
            'session_id': session_id,
            'pattern': pattern,
        })
    
    def auth(self) -> 'UserAuth':
        """获取用户认证API"""
        return UserAuth(self)
    
    def env(self) -> 'Environment':
        """获取环境检测API"""
        return Environment(self)


class UserAuth:
    """用户认证API"""
    
    def __init__(self, client: CaptchaClient):
        self.client = client
    
    def register(
        self,
        username: str,
        email: str,
        password: str,
        behavior_data: Optional[str] = None
    ) -> Dict:
        """
        用户注册
        
        Args:
            username: 用户名
            email: 邮箱
            password: 密码
            behavior_data: 行为数据
            
        Returns:
            注册结果
        """
        data = {
            'username': username,
            'email': email,
            'password': password,
        }
        if behavior_data:
            data['behavior_data'] = behavior_data
        
        return self.client._request('POST', '/api/v1/auth/register', data=data)
    
    def login(
        self,
        username: str,
        password: str,
        captcha_token: Optional[str] = None
    ) -> LoginResponse:
        """
        用户登录
        
        Args:
            username: 用户名
            password: 密码
            captcha_token: 验证码令牌
            
        Returns:
            登录响应
        """
        data = {'username': username, 'password': password}
        if captcha_token:
            data['captcha_token'] = captcha_token
        
        result = self.client._request('POST', '/api/v1/auth/login', data=data)
        self.client._token = result.get('access_token')
        self.client._refresh_token = result.get('refresh_token')
        return LoginResponse(**result)
    
    def refresh_token(self, refresh_token: Optional[str] = None) -> Dict:
        """
        刷新访问令牌
        
        Args:
            refresh_token: 刷新令牌
            
        Returns:
            刷新结果
        """
        token = refresh_token or self.client._refresh_token
        if not token:
            raise Exception("No refresh token available")
        
        result = self.client._request('POST', '/api/v1/auth/refresh', data={
            'refresh_token': token,
        })
        
        self.client._token = result.get('access_token')
        if result.get('refresh_token'):
            self.client._refresh_token = result.get('refresh_token')
        
        return result
    
    def logout(self) -> None:
        """用户登出"""
        try:
            self.client._request('POST', '/api/v1/auth/logout')
        finally:
            self.client._token = None
            self.client._refresh_token = None
    
    def verify_email(self, token: str) -> Dict:
        """
        验证邮箱
        
        Args:
            token: 验证令牌
            
        Returns:
            验证结果
        """
        return self.client._request('GET', '/api/v1/auth/verify-email', params={'token': token})
    
    def resend_verification(self, email: str) -> Dict:
        """
        重新发送验证邮件
        
        Args:
            email: 邮箱
            
        Returns:
            发送结果
        """
        return self.client._request('POST', '/api/v1/auth/resend-verification', data={'email': email})
    
    def request_password_reset(self, email: str) -> Dict:
        """
        请求重置密码
        
        Args:
            email: 邮箱
            
        Returns:
            请求结果
        """
        return self.client._request('POST', '/api/v1/auth/request-password-reset', data={'email': email})
    
    def reset_password(self, token: str, new_password: str) -> Dict:
        """
        重置密码
        
        Args:
            token: 重置令牌
            new_password: 新密码
            
        Returns:
            重置结果
        """
        return self.client._request('POST', '/api/v1/auth/reset-password', data={
            'token': token,
            'new_password': new_password,
        })


class Environment:
    """环境检测API"""
    
    def __init__(self, client: CaptchaClient):
        self.client = client
    
    def get_detection_script(self, callback: Optional[str] = None) -> str:
        """
        获取检测脚本
        
        Args:
            callback: 回调函数名
            
        Returns:
            脚本内容
        """
        params = {}
        if callback:
            params['callback'] = callback
        
        url = f"{self.client.base_url}/api/v1/detect/script"
        response = self.client.session.get(url, params=params, timeout=self.client.timeout)
        response.raise_for_status()
        return response.text
    
    def submit_detection(self, data: Dict) -> Dict:
        """
        提交检测数据
        
        Args:
            data: 检测数据
            
        Returns:
            提交结果
        """
        return self.client._request('POST', '/api/v1/detect/submit', data=data)
    
    def check_environment(self, data: Dict) -> Dict:
        """
        环境检测
        
        Args:
            data: 检测数据
            
        Returns:
            检测结果
        """
        return self.client._request('POST', '/api/v1/detect/check', data=data)


# 使用示例
def example_slider_captcha():
    """滑块验证码示例"""
    client = CaptchaClient('http://localhost:8080')
    
    # 获取验证码
    captcha = client.get_slider_captcha(width=320, height=160, tolerance=8)
    print(f"Session ID: {captcha.session_id}")
    
    # 生成模拟轨迹
    trajectory = [
        TrajectoryPoint(x=0, y=captcha.secret_y, t=0),
        TrajectoryPoint(x=50, y=captcha.secret_y + 5, t=200),
        TrajectoryPoint(x=100, y=captcha.secret_y - 3, t=400),
        TrajectoryPoint(x=150, y=captcha.secret_y + 2, t=600),
        TrajectoryPoint(x=185, y=captcha.secret_y, t=800),
    ]
    
    # 验证
    result = client.verify_captcha(
        session_id=captcha.session_id,
        x=185,
        y=captcha.secret_y,
        trajectory=trajectory
    )
    
    print(f"Success: {result.success}, Message: {result.message}")


def example_user_login():
    """用户登录示例"""
    client = CaptchaClient('http://localhost:8080')
    auth = client.auth()
    
    try:
        login_result = auth.login(username='testuser', password='password123')
        print(f"Login successful, access token: {login_result.access_token[:50}...")
    except Exception as e:
        print(f"Login failed: {e}")


if __name__ == '__main__':
    print("行为验证系统 Python SDK 示例")
    print("1. 滑块验证码示例")
    print("2. 用户登录示例")
    
    choice = input("请选择示例 (1/2): ").strip()
    
    if choice == '1':
        example_slider_captcha()
    elif choice == '2':
        example_user_login()
    else:
        print("无效的选择")
